includes/class-indieauth-token-endpoint.php

```diff
@@ -119,16 +122,47 @@ public function get( $request ) {
 		return rest_ensure_response( $token );
 	}
 
-	public function set_token( $token, $expiration = null ) {
+	public function set_token( $token, $expiration = null, $user_id = null ) {
 		if ( ! isset( $token['me'] ) ) {
 			return false;
 		}
-		$user = get_user_by_identifier( $token['me'] );
-		if ( $user instanceof WP_User ) {
-			$this->tokens->set_user( $user->ID );
-			return $this->tokens->set( $token, $expiration );
+		if ( ! $user_id ) {
+			$user_id = get_user_by_identifier( $token['me'] );
+			if ( $user instanceof WP_User ) {
+				$user_id = $user_id->ID;
+			} else {
+				return false;
+			}
 		}
-		return false;
+
+		$this->tokens->set_user( $user_id );
+
+		return $this->tokens->set( $token, $expiration );
+	}
+
+	/*
+	 * Sets a refresh token based on an access token.
+	 * @param array $token Access Token Return.
+	 * @param int $user User ID.
+	 * @return string Refresh Token.
+	 */
+	public function set_refresh_token( $token, $user ) {
+		$refresh = array(
+			'scope'     => $token['scope'],
+			'client_id' => $token['client_id'],
+			'iat'       => time(),
```

Comment:
looks like we don't use `exp` as the expiry is stored separately?

Reply:
It's added when the token is saved.